marlowe-playground-client/src/Blockly/Dom.purs

```diff
@@ -0,0 +1,253 @@
+  -- are the parsed <next> elements which represent the siblings of a parent Statement.
+  readAsBlock :: forall m. MonadEffect m => MonadThrow ReadDomError m => Element -> m (Tuple Block (Array Block))
+  readAsBlock element =
+    if Element.tagName element /= "block" then
+      throwError $ TypeMismatch element "block"
+    else do
+      blockId <- liftEffect $ Element.id element
+      blockType <- getAttribute "type" element
+      elementChildren <- getChildrens element
+      -- Parse each child element, which can be Either a BlockChild (field, value, statement) or a list of
+      -- siblings to pass to my parent
+      { left, right } <- separate <$> traverse readAsBlockChild elementChildren
+      let
+        children = Object.fromFoldable left
+      pure
+        $ Tuple
+            (Block { id: blockId, type: blockType, children })
+            (join right)
+
+  -- Tries to read an Element as the children of a Block node. The Left hand side of the return represents a Block direct children
+  -- while the Right hand side represents the parsed `<next>` elements which should be appended to a parents Statement.
+  readAsBlockChild ::
+    forall m.
+    MonadEffect m =>
+    MonadThrow ReadDomError m =>
+    Element ->
+    m (Either ChildWithName (Array Block))
+  readAsBlockChild element = do
+    case Element.tagName element of
+      "field" -> do
+        name <- getAttribute "name" element
+        value <- getElementText element
+        pure $ Left $ Tuple name $ Field value
+      "statement" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Left $ Tuple name $ Statement ([ block ] <> nexts)
+      "value" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        -- Even if a value can have a Block appended, it should have a single block child and no "siblings"
+        if length nexts /= 0 then
+          throwError $ IncorrectSiblingNesting "value"
+        else
+          pure $ Left $ Tuple name $ Value block
+      "next" -> do
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Right ([ block ] <> nexts)
+      _ -> throwError $ TypeMismatch element "field, statement, value, next"
+
+  getSingleChild :: forall m. MonadEffect m => MonadThrow ReadDomError m => Element -> m Element
+  getSingleChild element = do
+    children <- getChildrens element
+    if length children /= 1 then
+      throwError $ SingleChildExpected element $ length children
+    else do
+      pure $ unsafePartial $ UnsafeArray.head children
+
+  getChildrens :: forall m. MonadEffect m => Element -> m (Array Element)
+  getChildrens element =
+    liftEffect do
+      elements <- (ParentNode.children <<< Element.toParentNode) element
+      HTMLCollection.toArray elements
```

Comment:
what about
`liftEffect $ HTMLCollection.toArray <$> (ParentNode.children <<< Element.toParentNode) element`
?

Reply:
or in this case I might be tempted to change it to a concrete `Effect`
```
getChildren :: Element -> Effect (Array Element)
getChildren = HTMLCollection.toArray <$> (ParentNode.children <<< Element.toParentNode)
```

Reply:
I don't think a function that has a single action of a specific monad makes sense to have a class version of that monad. Same applies to others.

Reply:
But this are all helper functions inside `getDom`, the idea is to put the class version in the definition so the code that uses it is clearer. Will change to the `<$>` version

Reply:
well at the moment `getDom` is a concrete `Effect` (although I recommended not). So if you change to `getDom : MonadEffect m` then you would need `liftEffect` in many places in `getDom` but probably no more than you have here. I just don't think it makes sense to move away from the concrete `Effect` for a function of one line that does only mapping and composition of `Effect` functions

Reply:
actually couldn't use `<$>` as I needed bind. Point free version would be like 

```purescript
  getChildren :: Element -> Effect (Array Element)
  getChildren = HTMLCollection.toArray <=< ParentNode.children <<< Element.toParentNode
```

a more human friendly version is 

```purescript
  getChildren :: Element -> Effect (Array Element)
  getChildren element = HTMLCollection.toArray =<< (ParentNode.children $ Element.toParentNode element)
```

I went for the second one, but it's a matter of taste

Reply:
ooh, interesting. yes, I prefer the first one, it reads to me "get a parent node, then get it's parent, then turn it into an array" I'm not concerned with the 'implementation detail' of `<<<` verses `<=<` as I just want to know more or less what the function does.